### PR TITLE
Speed up property accesses using Symbols

### PR DESCRIPTION
**Pull request by sethbrenith**
This change puts a polymorphic inline cache on JavascriptSymbol so that it can act like PropertyString and cache information about loads and stores that use it. It also adds a cache mapping property IDs to JavascriptSymbol weak refs in ScriptContext, similar to the existing cache for PropertyStrings, and modifies existing code to not create new JavascriptSymbol instances unless necessary. When lowering a load or store with a likely symbol, we can emit code that is very similar to the fast path for strings.

### CODE REVIEW

File: lib/Runtime/Types/RecyclableObject.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace Js {
         InlineCache* inlineCache;
         PolymorphicInlineCache * polymorphicInlineCache;
         FunctionBody * functionBody;
-        PropertyString * propertyString;
+        PropertyRecordUsageCache* propertyRecordUsageCache; // Interior pointer to a Symbol or PropertyString. Okay as long as PropertyValueInfo is stack-allocated.
```
Comment (sethbrenith) [inline lib/Runtime/Types/RecyclableObject.h]: >Interior pointer [](start = 63, length = 16)

This seems moderately creepy; I'm open to other suggestions for how to represent a thing that's either a Symbol or a PropertyString.

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -15167,6 +15169,19 @@ Lowerer::GenerateFastElemICommon(
                 return nullptr;
             }
         }
+        else if (indexOpnd->GetValueType().IsLikelySymbol())
+        {
+            if (!baseOpnd->GetValueType().IsLikelyOptimizedTypedArray())
+            {
+                // If profile data says that it's a typed array - do not generate the symbol fast path as the src. could be a temp and that would cause a bug.
```
Comment (sethbrenith) [inline lib/Backend/Lower.cpp]: >could be a temp and that would cause a bug. [](start = 115, length = 43)

I copied this but don't really know if it's necessary in this case or even still relevant. Comment from change that introduced it is below. I removing the typed array checks in the string case above and running typedArrayProfile.js, and I saw no problems.

   Change 1113228 by REDMOND\nikhilkh@IEDRTHOST5-fbliedev1 on 2012/06/16 17:30:09
   
    WIN8:915315: [WebCrawler] [x86chk] [DynamicProfileMutator]:
    INVALID_POINTER_READ jscript9!Js::Type::GetTypeId+c
    [d:\work\enlist\fbl_ie_dev1\src\inetcore\
    The problem is that the PropertyString fast path was getting generated
    within a forin even though the profile data said that the store was on a
    typed array. Using the profile data, the src would be marked temp as it
    was a typed array store. At runtime, the property string fast path was
    actually be successful and store a temp stack allocated javascript
    number.
    When the stack allocated number was used, it would cause an AV.
    I have also added a unit test for this issue.
    IECAT #: 46183
    Code Reviewer: louisl
    Buddy Tester: saagarwa;sansun

File: lib/Runtime/Library/JavascriptObject.cpp [hunk 3]
```diff
@@ -1997,7 +1997,7 @@ namespace Js
                 // Also, if the object's type has not changed, we need to ensure that
                 // the cached property string for this property, if any, does not
                 // specify this object's type.
-                scriptContext->InvalidatePropertyStringCache(propId, obj->GetType());
+                scriptContext->InvalidatePropertyStringAndSymbolCaches(propId, obj->GetType());
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptObject.cpp]: I still haven't found how to construct a test case where there's actually work to do here. Either the writable data detection bit is not set, the type was replaced with a new type, or the property string was not cached. To test that the cache was being invalidated, I temporarily removed the writable-data-only check.

File: lib/Runtime/Types/RecyclableObject.h (hunk 1 above)
Comment (digitalinfinity) [reply lib/Runtime/Types/RecyclableObject.h]: What keeps the Symbol/PropertyString alive? You might want either an additional pointer to the DynamicObject that owns the PropertyRecordUsageCache or make this a union of Symbol & PropertyString?

---
In reply to: [162213674](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162213674) [](ancestors = 162213674)

File: lib/Runtime/Types/DynamicObjectPropertyEnumerator.cpp [hunk 4]
```diff
@@ -255,13 +255,18 @@ namespace Js
             }
         } while (Js::IsInternalPropertyId(propertyId));
 
-        if (info.GetPropertyString() != nullptr && info.GetPropertyString()->ShouldUseCache() && propertyString == info.GetPropertyString())
+        if (info.GetPropertyRecordUsageCache() != nullptr && info.GetPropertyRecordUsageCache()->ShouldUseCache())
         {
-            CacheOperators::CachePropertyRead(startingObject, this->object, false, propertyId, false, &info, scriptContext);
-            if ((!(this->flags & EnumeratorFlags::EphemeralReference)) && info.IsStoreFieldCacheEnabled() && info.IsWritable() && ((info.GetFlags() & (InlineCacheGetterFlag | InlineCacheSetterFlag)) == 0))
+            PropertyString* propertyStringAsPropertyString = propertyString == nullptr ? nullptr : PropertyString::TryFromVar(propertyString);
```
Comment (digitalinfinity) [inline lib/Runtime/Types/DynamicObjectPropertyEnumerator.cpp]: When would this be null if we have a cache and we should be using it?

File: lib/Runtime/Library/RuntimeLibraryPch.h [hunk 5]
```diff
@@ -39,7 +39,6 @@
 
 #include "Library/JavascriptVariantDate.h"
 #include "Library/JavascriptPromise.h"
-#include "Library/JavascriptSymbol.h"
```
Comment (digitalinfinity) [inline lib/Runtime/Library/RuntimeLibraryPch.h]: Why was this removed?

File: lib/Runtime/Library/PropertyRecordUsageCache.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        // Required due to DEFINE_VTABLE_CTOR on parent, but useless
+    }
+
+    PropertyRecordUsageCache::PropertyRecordUsageCache(StaticType* type, const PropertyRecord* propertyRecord) :
+        propertyRecord(propertyRecord),
+        ldElemInlineCache(nullptr),
+        stElemInlineCache(nullptr),
+        hitRate(0)
+    {
+        // TODO: in future, might be worth putting these inline to avoid extra allocations. PIC copy API needs to be updated to support this though
```
Comment (digitalinfinity) [inline lib/Runtime/Library/PropertyRecordUsageCache.cpp]: Actually, given this comment, the solution might be to allocate PropertyRecordUsageCache dynamically rather than statically? If you do that and then put the inline caches inline on it, you end up net-saving one allocation :)

Comment (digitalinfinity) [reply lib/Runtime/Library/PropertyRecordUsageCache.cpp]: Although, yes, you'd still have to be careful doing that since now your inline caches become interior pointers 

---
In reply to: [162229422](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162229422) [](ancestors = 162229422)

File: lib/Runtime/Types/DynamicObjectPropertyEnumerator.cpp (hunk 4 above)
Comment (sethbrenith) [reply lib/Runtime/Types/DynamicObjectPropertyEnumerator.cpp]: It looked to me like propertyString could be null if the flags said to fetch Symbol properties, though we may never use this method with those flags.

---
In reply to: [162228454](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162228454) [](ancestors = 162228454)

File: lib/Runtime/Types/RecyclableObject.h (hunk 1 above)
Comment (sethbrenith) [reply lib/Runtime/Types/RecyclableObject.h]: I like your idea of keeping an extra pointer to the owning object, thanks! It seems that PropertyValueInfo is always stack-allocated, and you mentioned that stack references are checked for interior pointers, but relying on that behavior feels very brittle.

---
In reply to: [162227539](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162227539) [](ancestors = 162227539,162213674)

File: lib/Runtime/Library/PropertyRecordUsageCache.cpp (hunk 6 above)
Comment (sethbrenith) [reply lib/Runtime/Library/PropertyRecordUsageCache.cpp]: I'd prefer to leave this alone for now (this comment was just copied over from PropertyString, along with most of this class's implementation).

---
In reply to: [162229510](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162229510) [](ancestors = 162229510,162229422)

**Comment by jackhorton** [comment]
If this is for Node, shouldn't it target release/1.9? Also should it get mainlined before 1.9.0/8.9.4 is cut?

**Comment by dilijev** [comment]
Squash is best-effort. If there's useful history here for posterity (changes that can be grouped by broad feature or change, or which makes certain changes easier to review), then keep some commits. In general, cleanup and merge commits are not useful. IMO this PR can be squashed to a single commit on top of the head of the branch.

File: lib/Runtime/Library/Chakra.Runtime.Library.vcxproj [hunk 7]
```diff
@@ -172,13 +172,15 @@
     <ClCompile Include="$(MSBuildThisFileDirectory)JSONStringBuilder.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)JSONStringifier.cpp" />
     <ClCompile Include="$(MSBuildThisFileDirectory)LazyJSONString.cpp" />
+    <ClCompile Include="PropertyRecordUsageCache.cpp" />
```
Comment (MikeHolman) [inline lib/Runtime/Library/Chakra.Runtime.Library.vcxproj]: nit: for razzle build to work, we need to prepend with $(MSBuildThisFileDirectory)

Comment (sethbrenith) [reply lib/Runtime/Library/Chakra.Runtime.Library.vcxproj]: Great catch! I just let VS do whatever it wanted to here, but it didn't match the correct style.

---
In reply to: [163133827](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r163133827) [](ancestors = 163133827)

**Comment by sethbrenith** [comment]
Okay, thanks. I'll squash before merging. None of these intermediate states are useful.

---
In reply to: [359649425](https://github.com/Microsoft/ChakraCore/pull/4569#issuecomment-359649425) [](ancestors = 359649425)

**Comment by sethbrenith** [comment]
Discussed offline; master is fine for this change.

---
In reply to: [359122247](https://github.com/Microsoft/ChakraCore/pull/4569#issuecomment-359122247) [](ancestors = 359122247)
